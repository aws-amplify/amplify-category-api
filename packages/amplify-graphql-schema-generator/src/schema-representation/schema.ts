import { Model } from './types';
import { Engine } from './engine';

export class Schema {
  private models: Model[] = [];
  constructor(private engine: Engine) {}

  public getModels(): Model[] {
    return this.models;
  }

  public getEngine(): Engine {
    return this.engine;
  }

  public addModel(model: Model): void {
    const modelName = model.getName();
    if (this.hasModel(modelName)) {
      throw new Error(`Model "${modelName}" already exists`);
    }
    this.models.push(model);
  }

  public hasModel(name: string): boolean {
    return this.models.some((model) => model.getName() === name);
  }
}
