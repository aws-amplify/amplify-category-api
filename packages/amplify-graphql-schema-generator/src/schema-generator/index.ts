export * from './generate-schema';
export * from './schema-overrides';
