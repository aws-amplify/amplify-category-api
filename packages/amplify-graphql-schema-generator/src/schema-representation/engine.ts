export type DBEngineType = 'MySQL' | 'Postgres' | 'DynamoDB';

export class Engine {
  constructor(public type: DBEngineType) {}
}
