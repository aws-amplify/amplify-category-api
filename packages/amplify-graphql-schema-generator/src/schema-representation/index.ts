export * from './engine';
export * from './schema';
export * from './types';
