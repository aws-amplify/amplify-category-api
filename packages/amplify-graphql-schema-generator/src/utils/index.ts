export * from './vpc-helper';
export { getSSLConfig } from './cert';
