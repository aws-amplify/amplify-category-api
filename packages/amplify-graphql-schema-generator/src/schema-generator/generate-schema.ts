import { DirectiveWrapper, EnumWrapper, FieldWrapper, ObjectDefinitionWrapper } from '@aws-amplify/graphql-transformer-core';
import {
  EnumValueDefinitionNode,
  Kind,
  print,
  DocumentNode,
  InputObjectTypeDefinitionNode,
  ListValueNode,
  StringValueNode,
  DirectiveNode,
  ObjectTypeDefinitionNode,
} from 'graphql';
import { EnumType, Field, FieldType, Index, Model, Schema } from '../schema-representation';
import { applySchemaOverrides } from './schema-overrides';
import { singular } from 'pluralize';
import { toCamelCase, toPascalCase } from 'graphql-transformer-common';

export const generateGraphQLSchema = (schema: Schema, existingSchemaDocument?: DocumentNode | undefined): string => {
  const models = schema.getModels();
  const document: any = {
    kind: Kind.DOCUMENT,
    definitions: [],
  };

  const { includeTables, excludeTables } = getIncludeExcludeConfig(existingSchemaDocument);

  models.forEach((model) => {
    // Verify whether a table should be included or excluded
    if (includeTables.length > 0 && !includeTables.includes(model.getName())) {
      return;
    }
    if (excludeTables.length > 0 && excludeTables.includes(model.getName())) {
      return;
    }

    const primaryKey = model.getPrimaryKey();
    if (!primaryKey) {
      return;
    }

    const type = constructObjectType(model);
    const fields = model.getFields();
    const primaryKeyFields = primaryKey?.getFields();
    fields.forEach((f) => {
      if (isEnum(f.type)) {
        const enumType = constructEnumType(getBaseType(f.type) as EnumType);
        if (!document.definitions.find((d) => d.name.value === enumType.name)) {
          document.definitions.push(enumType.serialize());
        }
      }

      const field: any = convertInternalFieldTypeToGraphQL(f, primaryKeyFields.includes(f.name));
      type.fields.push(field);
    });

    addPrimaryKey(type, model.getPrimaryKey());
    addIndexes(type, model.getIndexes());

    document.definitions.push(type.serialize());
  });

  const documentWithOverrides = applySchemaOverrides(document as DocumentNode, existingSchemaDocument);
  const schemaStr = printSchema(documentWithOverrides);
  return schemaStr;
};

const isEnum = (type: FieldType): boolean => {
  if (type.kind === 'NonNull' || type.kind === 'List') {
    return isEnum(type.type);
  }
  return type.kind === 'Enum';
};

const getBaseType = (type: FieldType): FieldType => {
  if (type.kind === 'NonNull' || type.kind === 'List') {
    return getBaseType(type.type);
  }
  return type;
};

const convertInternalFieldTypeToGraphQL = (field: Field, isPrimaryKeyField: boolean): FieldWrapper => {
  const fieldName = field.name;
  const typeWrappers = [];
  let fieldType = field.type;
  while (fieldType.kind !== 'Scalar' && fieldType.kind !== 'Custom' && fieldType.kind !== 'Enum') {
    typeWrappers.push(fieldType.kind);
    fieldType = (fieldType as any).type;
  }

  const fieldDirectives = [];

  // construct the refersTo directive
  const fieldTypeName = convertToGraphQLFieldName(fieldName);
  const fieldNameNeedsMapping = fieldTypeName !== fieldName;
  if (fieldNameNeedsMapping) {
    const fieldNameMappingDirective = getRefersToDirective(fieldName);
    fieldDirectives.push(fieldNameMappingDirective);
  }

  // construct the default directive
  const fieldHasDefaultValue = field?.default && field?.default?.value;
  const fieldIsOptional = fieldHasDefaultValue && !isPrimaryKeyField;
  if (fieldHasDefaultValue) {
    const defaultStringValue = String(field.default.value);
    if (!isComputeExpression(defaultStringValue)) {
      fieldDirectives.push(
        new DirectiveWrapper({
          kind: Kind.DIRECTIVE,
          name: {
            kind: 'Name',
            value: 'default',
          },
          arguments: [
            {
              kind: 'Argument',
              name: {
                kind: 'Name',
                value: 'value',
              },
              value: {
                kind: 'StringValue',
                value: defaultStringValue,
              },
            },
          ],
        }),
      );
    }
  }

  // Construct the field wrapper object
  const result = new FieldWrapper({
    kind: 'FieldDefinition',
    name: {
      kind: 'Name',
      value: fieldTypeName,
    },
    type: {
      kind: 'NamedType',
      name: {
        kind: 'Name',
        value: fieldType.name,
      },
    },
    directives: fieldDirectives,
  });

  while (typeWrappers.length > 0) {
    const wrapperType = typeWrappers.pop();
    if (wrapperType === 'List') {
      result.wrapListType();
    } else if (wrapperType === 'NonNull' && !fieldIsOptional) {
      result.makeNonNullable();
    }
  }

  return result;
};

const constructObjectType = (model: Model) => {
  const modelName = model.getName();
  const directives = [];

  const modelTypeName = convertToGraphQLTypeName(modelName);
  const modelNameNeedsMapping = modelTypeName !== modelName;
  if (modelNameNeedsMapping) {
    const modelNameMappingDirective = getRefersToDirective(modelName);
    directives.push(modelNameMappingDirective);
  }

  const modelDirective = {
    kind: Kind.DIRECTIVE,
    name: {
      kind: 'Name',
      value: 'model',
    },
  } as DirectiveNode;
  directives.push(modelDirective);

  return new ObjectDefinitionWrapper({
    kind: Kind.OBJECT_TYPE_DEFINITION,
    name: {
      kind: 'Name',
      value: modelTypeName,
    },
    fields: [],
    directives: directives,
  });
};

const constructEnumType = (type: EnumType): EnumWrapper => {
  const enumValues = type.values.map((t) => {
    return {
      kind: Kind.ENUM_VALUE_DEFINITION,
      name: {
        kind: 'Name',
        value: t,
      },
    } as EnumValueDefinitionNode;
  });
  const enumType = new EnumWrapper({
    kind: Kind.ENUM_TYPE_DEFINITION,
    name: {
      kind: 'Name',
      value: type.name,
    },
    values: enumValues,
  });
  return enumType;
};

const addIndexes = (type: ObjectDefinitionWrapper, indexes: Index[]): void => {
  indexes.forEach((index) => {
    const firstField = convertToGraphQLFieldName(index.getFields()[0]);
    const indexField = type.getField(firstField);

    const indexArguments = [];
    indexArguments.push({
      kind: 'Argument',
      name: {
        kind: 'Name',
        value: 'name',
      },
      value: {
        kind: 'StringValue',
        value: index.name,
      },
    });

    if (index.getFields().length > 1) {
      indexArguments.push({
        kind: 'Argument',
        name: {
          kind: 'Name',
          value: 'sortKeyFields',
        },
        value: {
          kind: 'ListValue',
          values: index
            .getFields()
            .slice(1)
            .map((k) => {
              return {
                kind: 'StringValue',
                value: convertToGraphQLFieldName(k),
              };
            }),
        },
      });
    }

    indexField.directives.push(
      new DirectiveWrapper({
        kind: Kind.DIRECTIVE,
        name: {
          kind: 'Name',
          value: 'index',
        },
        arguments: indexArguments,
      }),
    );
  });
};

const addPrimaryKey = (type: ObjectDefinitionWrapper, primaryKey: Index): void => {
  if (!primaryKey) {
    return;
  }

  const firstField = convertToGraphQLFieldName(primaryKey.getFields()[0]);
  const primaryKeyField = type.getField(firstField);
  const keyArguments = [];

  if (primaryKey.getFields().length > 1) {
    keyArguments.push({
      kind: 'Argument',
      name: {
        kind: 'Name',
        value: 'sortKeyFields',
      },
      value: {
        kind: 'ListValue',
        values: primaryKey
          .getFields()
          .slice(1)
          .map((k) => {
            return {
              kind: 'StringValue',
              value: convertToGraphQLFieldName(k),
            };
          }),
      },
    });
  }

  primaryKeyField.directives.push(
    new DirectiveWrapper({
      kind: Kind.DIRECTIVE,
      name: {
        kind: 'Name',
        value: 'primaryKey',
      },
      arguments: keyArguments,
    }),
  );
};

type IncludeExcludeConfig = {
  includeTables: string[];
  excludeTables: string[];
};

const getIncludeExcludeConfig = (document: DocumentNode | undefined): IncludeExcludeConfig => {
  const emptyConfig = { includeTables: [], excludeTables: [] };
  if (!document) {
    return emptyConfig;
  }

  const amplifyInputType = document.definitions.find(
    (d) => d.kind === 'InputObjectTypeDefinition' && d.name.value === 'AMPLIFY',
  ) as InputObjectTypeDefinitionNode;
  if (!amplifyInputType) {
    return emptyConfig;
  }

  const includeFieldNodeValue = amplifyInputType.fields.find((f) => f.name.value === 'include')?.defaultValue;
  const excludeFieldNodeValue = amplifyInputType.fields.find((f) => f.name.value === 'exclude')?.defaultValue;

  if (includeFieldNodeValue && includeFieldNodeValue.kind !== 'ListValue') {
    throw new Error('Invalid value for include option. Please check your GraphQL schema.');
  }

  if (excludeFieldNodeValue && excludeFieldNodeValue.kind !== 'ListValue') {
    throw new Error('Invalid value for include option. Please check your GraphQL schema.');
  }

  const includeTables = (includeFieldNodeValue as ListValueNode)?.values.map((v: StringValueNode) => v.value);
  const excludeTables = (excludeFieldNodeValue as ListValueNode)?.values.map((v: StringValueNode) => v.value);
  if (includeTables && includeTables.length > 0 && excludeTables && excludeTables.length > 0) {
    throw new Error('Cannot specify both include and exclude options. Please check your GraphQL schema.');
  }

  return {
    includeTables: includeTables || [],
    excludeTables: excludeTables || [],
  };
};

/**
 * Checks if the user defined default for a field is a compute expression
 * @param value The default value for a field
 * @returns if the default value is a compute expression like for example (RAND() * RAND()))
 */
export const isComputeExpression = (value: string) => {
  /* As per MySQL 8.x, 
    Complex computed expression default values like "(RAND() * RAND())" are enclosed within parentheses.
    Simple computed expression default values like "RAND()" are not. 
    These functions could have one or more arguments too, like "COS(PI())".
  */
  const isSimpleComputedExpression = value.match(/^[a-zA-Z0-9]+\(.*\)/);
  const isComplexComputedExpression = value.match(/^\([a-zA-Z0-9]+\(.*\)\)/);
  if (isSimpleComputedExpression || isComplexComputedExpression) {
    return true;
  }
  return false;
};

export const getRefersToDirective = (name: string): DirectiveNode => {
  return {
    kind: Kind.DIRECTIVE,
    name: {
      kind: 'Name',
      value: 'refersTo',
    },
    arguments: [
      {
        kind: 'Argument',
        name: {
          kind: 'Name',
          value: 'name',
        },
        value: {
          kind: 'StringValue',
          value: name,
        },
      },
    ],
  } as DirectiveNode;
};

export const convertToGraphQLTypeName = (modelName: string): string => {
  const cleanedInput = cleanMappedName(modelName);

  // Convert to PascalCase and Singularize
  return singular(toPascalCase(cleanedInput?.split('_')));
};

export const convertToGraphQLFieldName = (fieldName: string): string => {
  const cleanedInput = cleanMappedName(fieldName, true);

  // Convert to camelCase
  return toCamelCase(cleanedInput?.split('_'));
};

const cleanMappedName = (name: string, isField: boolean = false): string => {
  // If it does not have any alphabetic characters, use a meaningful name
  if (!name?.match(/[a-zA-Z]/)) {
    const suffix = name?.replace(/[^0-9]+/g, '');
    return isField ? `field${suffix}` : `Model${suffix}`;
  }

  // Remove leading digits and non-alphabetic characters
  // Convert non-alphanumeric characters to underscores
  const cleanedInput = name
    .replace(/^[^a-zA-Z]+/, '')
    .replace(/[^a-zA-Z0-9_]+/g, '_')
    .trim();

  return cleanedInput;
};

export const printSchema = (document: DocumentNode): string => {
  const sortedDocument = sortDocument(document);
  return print(sortedDocument);
};

const sortDocument = (document: DocumentNode): DocumentNode => {
  const documentWrapper = document as any;
  documentWrapper.definitions = [...document?.definitions].sort((def1, def2) => {
    if ((def1 as ObjectTypeDefinitionNode)?.name?.value > (def2 as ObjectTypeDefinitionNode)?.name?.value) {
      return 1;
    }
    if ((def1 as ObjectTypeDefinitionNode)?.name?.value < (def2 as ObjectTypeDefinitionNode)?.name?.value) {
      return -1;
    }
    return 0;
  });
  return documentWrapper;
};
