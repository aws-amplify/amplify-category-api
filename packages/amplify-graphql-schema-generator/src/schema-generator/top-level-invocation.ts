import * as os from 'os';
import { ImportedRDSType, ImportedDataSourceConfig } from '@aws-amplify/graphql-transformer-core';
import { DataSourceAdapter, MySQLDataSourceAdapter, MySQLDataSourceConfig } from '../datasource-adapter';
import { Engine, Schema } from '../schema-representation';
import { generateGraphQLSchema } from './generate-schema';

type AmplifyInputEntry = {
  name: string;
  type: string;
  default: string | number;
  comment?: string | undefined;
};

export const generateRDSSchema = async (databaseConfig: ImportedDataSourceConfig): Promise<string> => {
  // Establish the connection
  let adapter: DataSourceAdapter;
  let schema: Schema;
  switch (databaseConfig.engine) {
    case ImportedRDSType.MYSQL:
      adapter = new MySQLDataSourceAdapter(databaseConfig as MySQLDataSourceConfig);
      schema = new Schema(new Engine('MySQL'));
      break;
    default:
      console.error('Only MySQL Data Source is supported.');
  }

  try {
    await adapter.initialize();
  } catch (error) {
    console.error(
      'Failed to connect to the specified RDS Data Source. Check the connection details in the schema and re-try. Use "amplify api update-secrets" to update the user credentials.',
    );
    throw error;
  }

  const models = await adapter.getModels();
  adapter.cleanup();
  models.forEach((m) => schema.addModel(m));

  const schemaString =
    (await constructDefaultGlobalAmplifyInput(databaseConfig.engine, false)) + os.EOL + os.EOL + generateGraphQLSchema(schema);
  return schemaString;
};

const getGlobalAmplifyInputEntries = async (
  dataSourceType = ImportedRDSType.MYSQL,
  includeAuthRule = true,
): Promise<AmplifyInputEntry[]> => {
  const inputs: AmplifyInputEntry[] = [
    {
      name: 'engine',
      type: 'String',
      default: dataSourceType,
    },
  ];

  if (includeAuthRule) {
    inputs.push({
      name: 'globalAuthRule',
      type: 'AuthRule',
      default: '{ allow: public }',
      comment: `This "input" configures a global authorization rule to enable public access to all models in this schema.
    Learn more about authorization rules here: https://docs.amplify.aws/lib/graphql/auth-rules/`,
    });
  }
  return inputs;
};

const constructDefaultGlobalAmplifyInput = async (dataSourceType: ImportedRDSType, includeAuthRule: boolean): Promise<string> => {
  const inputs = await getGlobalAmplifyInputEntries(dataSourceType, includeAuthRule);
  const inputsString = inputs.reduce(
    (acc: string, input): string =>
      acc +
      ` ${input.name}: ${input.type} = ${input.type === 'String' ? '"' + input.default + '"' : input.default} ${
        input.comment ? '# ' + input.comment : ''
      } \n`,
    '',
  );
  return `input Amplify {\n${inputsString}}\n`;
};
