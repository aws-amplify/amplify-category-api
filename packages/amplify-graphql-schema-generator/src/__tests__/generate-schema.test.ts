import path from 'path';
import * as fs from 'fs';
import { parse } from 'graphql';
import { convertToGraphQLTypeName, printSchema, convertToGraphQLFieldName } from '../schema-generator/generate-schema';
import { Engine, Field, Model, Schema } from '../schema-representation';
import { generateGraphQLSchema } from '../schema-generator';
import { getSSLConfig } from '../utils';

describe('Type name conversions', () => {
  it('GraphQL idiomatic type name conversions', () => {
    // Singularize and Pascal case
    expect(convertToGraphQLTypeName('posts')).toEqual('Post');
    // Singularize
    expect(convertToGraphQLTypeName('Salaries')).toEqual('Salary');
    expect(convertToGraphQLTypeName('Lotuses')).toEqual('Lotus');
    // Pascal case
    expect(convertToGraphQLTypeName('employees_salaries')).toEqual('EmployeesSalary');

    // Remove special characters not supported in GraphQL
    expect(convertToGraphQLTypeName('Employees_salaries-Tables')).toEqual('EmployeesSalariesTable');
    expect(convertToGraphQLTypeName('_employee_Salaries-tables')).toEqual('EmployeeSalariesTable');
    expect(convertToGraphQLTypeName('Employees$salaries-#Table%log@Rates!types')).toEqual('EmployeesSalariesTableLogRatesType');

    // Remove numeric or special character prefix
    expect(convertToGraphQLTypeName('1Employee')).toEqual('Employee');
    expect(convertToGraphQLTypeName('12_123Employee')).toEqual('Employee');
    expect(convertToGraphQLTypeName('_123Employee')).toEqual('Employee');
    expect(convertToGraphQLTypeName('-#123Employee')).toEqual('Employee');
    expect(convertToGraphQLTypeName('123-Employee_345')).toEqual('Employee345');
    expect(convertToGraphQLTypeName('123__Employee_345')).toEqual('Employee345');

    // If only non-alphabetic characters are present, use a meaningful name
    expect(convertToGraphQLTypeName('123')).toEqual('Model123');
    expect(convertToGraphQLTypeName('_123')).toEqual('Model123');
    expect(convertToGraphQLTypeName('_#123')).toEqual('Model123');
  });

  it('infers refersTo from model name', () => {
    const dbschema = new Schema(new Engine('MySQL'));
    let model = new Model('users');
    model.addField(new Field('id', { kind: 'NonNull', type: { kind: 'Scalar', name: 'String' } }));
    model.addField(new Field('name', { kind: 'Scalar', name: 'String' }));
    model.setPrimaryKey(['id']);
    dbschema.addModel(model);

    model = new Model('profile');
    model.addField(new Field('id', { kind: 'NonNull', type: { kind: 'Scalar', name: 'String' } }));
    model.addField(new Field('details', { kind: 'Scalar', name: 'String' }));
    model.setPrimaryKey(['id']);
    dbschema.addModel(model);

    const graphqlSchema = generateGraphQLSchema(dbschema);
    expect(graphqlSchema).toMatchSnapshot();
  });
});

describe('enum imports', () => {
  it('generates enum imports', () => {
    const dbschema = new Schema(new Engine('Postgres'));
    const model = new Model('users');
    model.addField(new Field('id', { kind: 'NonNull', type: { kind: 'Scalar', name: 'String' } }));
    model.addField(new Field('name', { kind: 'Scalar', name: 'String' }));
    model.addField(new Field('status', { kind: 'Enum', name: 'UserStatus', values: ['ACTIVE', 'INACTIVE'] }));
    model.setPrimaryKey(['id']);
    dbschema.addModel(model);

    const graphqlSchema = generateGraphQLSchema(dbschema);
    expect(graphqlSchema).toMatchSnapshot();
  });

  it('enum values with single character', () => {
    const dbschema = new Schema(new Engine('Postgres'));
    const model = new Model('users');
    model.addField(new Field('id', { kind: 'NonNull', type: { kind: 'Scalar', name: 'String' } }));
    model.addField(new Field('name', { kind: 'Scalar', name: 'String' }));
    model.addField(new Field('status', { kind: 'Enum', name: 'UserStatus', values: ['A', 'D', 'B'] }));
    model.setPrimaryKey(['id']);
    dbschema.addModel(model);

    const graphqlSchema = generateGraphQLSchema(dbschema);
    parse(graphqlSchema); // Parse to make sure that the Schema is valid.
    expect(graphqlSchema).toMatchSnapshot();
  });

  it('invalid enum value must throw an error', () => {
    const dbschema = new Schema(new Engine('Postgres'));
    const model = new Model('users');
    model.addField(new Field('id', { kind: 'NonNull', type: { kind: 'Scalar', name: 'String' } }));
    model.addField(new Field('name', { kind: 'Scalar', name: 'String' }));
    model.addField(new Field('status', { kind: 'Enum', name: 'UserStatus', values: ['ACTIVE-0', 'INACTIVE-1'] }));
    model.setPrimaryKey(['id']);
    dbschema.addModel(model);

    expect(() => generateGraphQLSchema(dbschema)).toThrowError(
      'Enum "UserStatus" (values: ACTIVE-0,INACTIVE-1) contains one or more invalid values. Enum values must match the regex [_A-Za-z][_0-9A-Za-z]*.',
    );
  });

  it('multiple enum fields of same name must generate one graphql enum type', () => {
    const dbschema = new Schema(new Engine('Postgres'));
    let model = new Model('users');
    model.addField(new Field('id', { kind: 'NonNull', type: { kind: 'Scalar', name: 'String' } }));
    model.addField(new Field('name', { kind: 'Scalar', name: 'String' }));
    model.addField(new Field('status', { kind: 'Enum', name: 'UserStatus', values: ['ACTIVE', 'INACTIVE'] }));
    model.setPrimaryKey(['id']);
    dbschema.addModel(model);

    model = new Model('profile');
    model.addField(new Field('id', { kind: 'NonNull', type: { kind: 'Scalar', name: 'String' } }));
    model.addField(new Field('details', { kind: 'Scalar', name: 'String' }));
    model.addField(new Field('profilestatus', { kind: 'Enum', name: 'UserStatus', values: ['ACTIVE', 'INACTIVE'] }));
    model.setPrimaryKey(['id']);
    dbschema.addModel(model);

    const graphqlSchema = generateGraphQLSchema(dbschema);
    expect(graphqlSchema).toMatchSnapshot();
  });

  it('generates nonnull enum types correctly', () => {
    const dbschema = new Schema(new Engine('Postgres'));
    const model = new Model('users');
    model.addField(new Field('id', { kind: 'NonNull', type: { kind: 'Scalar', name: 'String' } }));
    model.addField(new Field('name', { kind: 'Scalar', name: 'String' }));
    model.addField(new Field('status', { kind: 'NonNull', type: { kind: 'Enum', name: 'UserStatus', values: ['ACTIVE', 'INACTIVE'] } }));
    model.setPrimaryKey(['id']);
    dbschema.addModel(model);

    const graphqlSchema = generateGraphQLSchema(dbschema);
    expect(graphqlSchema).toMatchSnapshot();
  });
});

describe('Field name conversions', () => {
  it('GraphQL idiomatic field name conversions', () => {
    // leave as-is
    expect(convertToGraphQLFieldName('posts')).toEqual('posts');
    expect(convertToGraphQLFieldName('postscolumn')).toEqual('postscolumn');
    // Camel case
    expect(convertToGraphQLFieldName('employees_salary')).toEqual('employeesSalary');

    // Remove special characters not supported in GraphQL
    expect(convertToGraphQLFieldName('Employees_salaries-Column')).toEqual('employeesSalariesColumn');
    expect(convertToGraphQLFieldName('_employee_Salaries-column')).toEqual('employeeSalariesColumn');
    expect(convertToGraphQLFieldName('Employees$salaries-#Table%log@Rates!types')).toEqual('employeesSalariesTableLogRatesTypes');
    expect(convertToGraphQLFieldName('ID')).toEqual('iD');
    expect(convertToGraphQLFieldName('MyID')).toEqual('myID');

    // Remove numeric or special character prefix
    expect(convertToGraphQLFieldName('1Employee')).toEqual('employee');
    expect(convertToGraphQLFieldName('12_123Employee')).toEqual('employee');
    expect(convertToGraphQLFieldName('_123Employee')).toEqual('employee');
    expect(convertToGraphQLFieldName('-#123Employee')).toEqual('employee');
    expect(convertToGraphQLFieldName('123-Employee_345')).toEqual('employee345');
    expect(convertToGraphQLFieldName('123__Employee_345')).toEqual('employee345');

    // If only non-alphabetic characters are present, use a meaningful name
    expect(convertToGraphQLFieldName('123')).toEqual('field123');
    expect(convertToGraphQLFieldName('_123')).toEqual('field123');
    expect(convertToGraphQLFieldName('_#123')).toEqual('field123');
  });

  it('infers refersTo from column names', () => {
    const dbschema = new Schema(new Engine('MySQL'));
    let model = new Model('users');
    model.addField(new Field('Id', { kind: 'NonNull', type: { kind: 'Scalar', name: 'String' } }));
    model.addField(new Field('name_field', { kind: 'Scalar', name: 'String' }));
    model.setPrimaryKey(['Id', 'name_field']);
    dbschema.addModel(model);

    model = new Model('profile');
    model.addField(new Field('id', { kind: 'NonNull', type: { kind: 'Scalar', name: 'String' } }));
    model.addField(new Field('Details', { kind: 'Scalar', name: 'String' }));
    model.setPrimaryKey(['id']);
    model.addIndex('profilesByDetails', ['Details']);
    dbschema.addModel(model);

    const graphqlSchema = generateGraphQLSchema(dbschema);
    expect(graphqlSchema).toMatchSnapshot();
  });
});

describe('Format generated schema', () => {
  it('generates schema with consistent type ordering', () => {
    const document = parse(`
        type User @model {
            id: ID!
        }
        type Profile @model {
            id: ID!
        }
    `);
    expect(printSchema(document)).toMatchInlineSnapshot(`
      "type Profile @model {
        id: ID!
      }

      type User @model {
        id: ID!
      }
      "
    `);
  });

  describe('getSSLConfig should use the correct certificate', () => {
    it('should use the bundled certificate', () => {
      const sslConfig = getSSLConfig('random-host@rds.amazonaws.com');
      const RDS_CERT_FILE_NAME = 'aws-rds-global-bundle.pem';
      const RDS_CERT_FILE_PATH = path.join(__dirname, '..', '..', 'certs', RDS_CERT_FILE_NAME);
      expect(sslConfig).toEqual(
        expect.objectContaining({
          rejectUnauthorized: true,
          ca: fs.readFileSync(RDS_CERT_FILE_PATH, 'utf-8'),
        }),
      );
    });

    it('should use the default trust store', () => {
      const sslConfig = getSSLConfig('random-host@db-provider.com');
      expect(sslConfig).toEqual(
        expect.objectContaining({
          rejectUnauthorized: true,
        }),
      );
    });

    it('should use the custom certificate', () => {
      const sslConfig = getSSLConfig('random-host@rexternal-db.com', 'SSL_CERTIFICATE');
      expect(sslConfig).toEqual(
        expect.objectContaining({
          rejectUnauthorized: true,
          ca: 'SSL_CERTIFICATE',
        }),
      );
    });
  });
});
