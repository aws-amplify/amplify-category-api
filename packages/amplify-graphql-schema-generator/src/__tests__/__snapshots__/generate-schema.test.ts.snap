// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Field name conversions infers refersTo from column names 1`] = `
"type Profile @refersTo(name: \\"profile\\") @model {
  id: String! @primaryKey
  details: String @refersTo(name: \\"Details\\") @index(name: \\"profilesByDetails\\")
}

type User @refersTo(name: \\"users\\") @model {
  id: String! @refersTo(name: \\"Id\\") @primaryKey(sortKeyFields: [\\"nameField\\"])
  nameField: String @refersTo(name: \\"name_field\\")
}
"
`;

exports[`Type name conversions infers refersTo from model name 1`] = `
"type Profile @refersTo(name: \\"profile\\") @model {
  id: String! @primaryKey
  details: String
}

type User @refersTo(name: \\"users\\") @model {
  id: String! @primaryKey
  name: String
}
"
`;

exports[`enum imports enum values with single character 1`] = `
"type User @refersTo(name: \\"users\\") @model {
  id: String! @primaryKey
  name: String
  status: UserStatus
}

enum UserStatus {
  A
  D
  B
}
"
`;

exports[`enum imports generates enum imports 1`] = `
"type User @refersTo(name: \\"users\\") @model {
  id: String! @primaryKey
  name: String
  status: UserStatus
}

enum UserStatus {
  ACTIVE
  INACTIVE
}
"
`;

exports[`enum imports generates nonnull enum types correctly 1`] = `
"type User @refersTo(name: \\"users\\") @model {
  id: String! @primaryKey
  name: String
  status: UserStatus!
}

enum UserStatus {
  ACTIVE
  INACTIVE
}
"
`;

exports[`enum imports multiple enum fields of same name must generate one graphql enum type 1`] = `
"type Profile @refersTo(name: \\"profile\\") @model {
  id: String! @primaryKey
  details: String
  profilestatus: UserStatus
}

type User @refersTo(name: \\"users\\") @model {
  id: String! @primaryKey
  name: String
  status: UserStatus
}

enum UserStatus {
  ACTIVE
  INACTIVE
}
"
`;
